File: app/src/main/java/com/baidu/tv/player/ui/main/PlaylistAdapter.java
package com.baidu.tv.player.ui.main;

import android.content.Context;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.baidu.tv.player.R;
import com.baidu.tv.player.model.Playlist;

import java.util.ArrayList;
import java.util.List;

/**
 * 播放列表适配器
 */
public class PlaylistAdapter extends RecyclerView.Adapter<PlaylistCardViewHolder> {
    private List<Playlist> playlists = new ArrayList<>();
    private OnItemClickListener onItemClickListener;
    private OnItemLongClickListener onItemLongClickListener;
    private OnDeleteClickListener onDeleteClickListener;
    private OnRefreshClickListener onRefreshClickListener;
    private final Context context;
    private boolean isEditMode = false;

    public PlaylistAdapter(Context context) {
        this.context = context;
    }

    public void setPlaylists(List<Playlist> playlists) {
        this.playlists = playlists;
        notifyDataSetChanged();
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.onItemClickListener = listener;
    }
    
    public void setOnItemLongClickListener(OnItemLongClickListener listener) {
        this.onItemLongClickListener = listener;
    }
    
    public void setOnDeleteClickListener(OnDeleteClickListener listener) {
        this.onDeleteClickListener = listener;
    }
    
    public void setOnRefreshClickListener(OnRefreshClickListener listener) {
        this.onRefreshClickListener = listener;
    }
    
    /**
     * 设置编辑模式
     */
    public void setEditMode(boolean editMode) {
        this.isEditMode = editMode;
        notifyDataSetChanged();
    }
    
    /**
     * 检查是否处于编辑模式
     */
    public boolean isEditMode() {
        return isEditMode;
    }

    @NonNull
    @Override
    public PlaylistCardViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_playlist_card, parent, false);
        return new PlaylistCardViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PlaylistCardViewHolder holder, int position) {
        Playlist playlist = playlists.get(position);
        
        // 格式化统计信息
        String stats;
        if (playlist.getTotalDuration() > 0) {
            String duration = DateUtils.formatElapsedTime(playlist.getTotalDuration() / 1000);
            stats = playlist.getTotalItems() + "个文件 · " + duration;
        } else {
            stats = playlist.getTotalItems() + "个文件";
        }
        
        holder.bind(playlist.getCoverImagePath(), playlist.getName(), stats);
        
        // 设置删除按钮和刷新按钮的可见性
        if (isEditMode) {
            holder.ivDelete.setVisibility(View.VISIBLE);
            holder.ivRefresh.setVisibility(View.VISIBLE);
        } else {
            holder.ivDelete.setVisibility(View.GONE);
            holder.ivRefresh.setVisibility(View.GONE);
        }
        
        // 点击事件
        holder.itemView.setOnClickListener(v -> {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(playlist);
            }
        });
        
        // 长按事件
        holder.itemView.setOnLongClickListener(v -> {
            if (onItemLongClickListener != null) {
                onItemLongClickListener.onItemLongClick(playlist);
                return true;
            }
            return false;
        });
        
        // 删除按钮点击事件
        holder.ivDelete.setOnClickListener(v -> {
            if (onDeleteClickListener != null) {
                onDeleteClickListener.onDeleteClick(playlist);
            }
        });
        
        // 刷新按钮点击事件
        holder.ivRefresh.setOnClickListener(v -> {
            if (onRefreshClickListener != null) {
                onRefreshClickListener.onRefreshClick(playlist);
            }
        });
    }

    @Override
    public int getItemCount() {
        return playlists.size();
    }

    public interface OnItemClickListener {
        void onItemClick(Playlist playlist);
    }
    
    public interface OnItemLongClickListener {
        void onItemLongClick(Playlist playlist);
    }
    
    public interface OnDeleteClickListener {
        void onDeleteClick(Playlist playlist);
    }
    
    public interface OnRefreshClickListener {
        void onRefreshClick(Playlist playlist);
    }
}

File: app/src/main/java/com/baidu/tv/player/kt/ui/main/PlaylistAdapter.kt
package com.baidu.tv.player.kt.ui.main

import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.baidu.tv.player.kt.databinding.ItemPlaylistCardBinding
import com.baidu.tv.player.kt.model.Playlist

class PlaylistAdapter(
    private val accessTokenProvider: () -> String? = { null },
) : RecyclerView.Adapter<PlaylistCardViewHolder>() {

    private val playlists = mutableListOf<Playlist>()
    private val refreshingPlaylistIds = mutableSetOf<Long>()

    var onItemClick: ((Playlist) -> Unit)? = null
    var onItemLongClick: ((Playlist) -> Unit)? = null
    var onDeleteClick: ((Playlist) -> Unit)? = null
    /** 卡片焦点变化回调：参数为当前获焦的播放列表，失去焦点时为 null。 */
    var onItemFocusChange: ((Playlist?) -> Unit)? = null

    /** 当前获焦卡片对应的播放列表，供菜单键等快捷操作定位目标。 */
    var focusedPlaylist: Playlist? = null
        private set

    var isEditMode: Boolean = false
        private set

    fun setPlaylists(newPlaylists: List<Playlist>) {
        playlists.clear()
        playlists.addAll(newPlaylists)
        notifyDataSetChanged()
    }

    fun setRefreshing(playlistId: Long, refreshing: Boolean) {
        if (refreshing) refreshingPlaylistIds += playlistId else refreshingPlaylistIds -= playlistId
        val position = playlists.indexOfFirst { it.id == playlistId }
        if (position >= 0) notifyItemChanged(position)
    }

    fun setEditMode(editMode: Boolean) {
        if (isEditMode == editMode) return
        isEditMode = editMode
        notifyDataSetChanged()
    }

    fun getItem(position: Int): Playlist = playlists[position]

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PlaylistCardViewHolder {
        val binding = ItemPlaylistCardBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return PlaylistCardViewHolder(binding, accessTokenProvider)
    }

    override fun onBindViewHolder(holder: PlaylistCardViewHolder, position: Int) {
        val playlist = playlists[position]
        val stats = if (playlist.totalDuration > 0) {
            "${playlist.totalItems}个文件 · ${DateUtils.formatElapsedTime(playlist.totalDuration / 1000)}"
        } else {
            "${playlist.totalItems}个文件"
        }
        holder.bind(playlist.coverImagePath, playlist.name, stats)

        holder.binding.ivDelete.visibility = if (isEditMode) View.VISIBLE else View.GONE
        updateRefreshIndicator(holder, playlist)
        holder.itemView.setOnFocusChangeListener { _, hasFocus ->
            focusedPlaylist = if (hasFocus) playlist else null
            onItemFocusChange?.invoke(focusedPlaylist)
            updateRefreshIndicator(holder, playlist)
        }

        holder.itemView.setOnClickListener { onItemClick?.invoke(playlist) }
        holder.itemView.setOnLongClickListener {
            // 刷新由卡片长按触发，避免 TV 焦点进入卡片内部；同步中禁止重复触发。
            if (playlist.id !in refreshingPlaylistIds) onItemLongClick?.invoke(playlist)
            true
        }
        holder.binding.ivDelete.setOnClickListener { onDeleteClick?.invoke(playlist) }
    }

    /**
     * 刷新指示：同步中显示转圈进度；未同步时刷新图标仅作「可长按同步」的视觉提示，
     * 焦点落在卡片上才浮现，避免常驻图标遮挡封面。
     */
    private fun updateRefreshIndicator(holder: PlaylistCardViewHolder, playlist: Playlist) {
        val isRefreshing = playlist.id in refreshingPlaylistIds
        holder.binding.progressRefresh.visibility = if (isRefreshing) View.VISIBLE else View.GONE
        holder.binding.ivRefresh.visibility =
            if (!isRefreshing && holder.itemView.isFocused) View.VISIBLE else View.GONE
    }

    override fun getItemCount(): Int = playlists.size
}
